### DISCUSSION TITLE
[Q] Writing large datasets - performance improvements possible?

### DISCUSSION THREAD

**Discussion by Tilogorn** [exceljs/exceljs#2449] [Q&A]
## 💬 Questions and Help

I'm writing a matrix of 500k rows multiplied by 6 columns to a workbook with excel.js on an embedded device (sensor data). This export takes more than two minutes.

For simulating the situation, see this mwe:

```js
const { performance } = require('perf_hooks');
var t0 = performance.now();

const Excel = require('exceljs');

// Generate fake data
var csv_data = [];
for (var i = 0; i < 500000; i++) {
    csv_data.push([i+1, i+2, i+3, i+4, i+5, i+6, i+7]);
}
console.log('Data generated: ' + (performance.now() - t0) + ' milliseconds.');

// Create XLSX
var workbook_write_stream = new Excel.stream.xlsx.WorkbookWriter({
    filename: './Test.xlsx',
    useStyles: true
});

var sheet = workbook_write_stream.addWorksheet('Data');
csv_data.forEach(function(line) {
    sheet.addRow(line).commit();
});

sheet.commit();

workbook_write_stream.commit()
    .then(function () {
        console.log('XLSX generated: ' + (performance.now() - t0) + ' milliseconds.');
    });
```
The output on my dev machine is
```
Data generated: 249.29447999969125 milliseconds.
XLSX generated: 17190.792275000364 milliseconds.
```
So this takes more than 17 seconds. Setting `useStyles` to `false` leads to these results:
```
Data generated: 245.52504999935627 milliseconds.
XLSX generated: 6420.7135119996965 milliseconds.
```
 This is a huge improvement but unfortunately "no styling" does not meet the requirements of the management ;-)

I need to `stream` and to `commit` because the RAM is limited on the embedded device. Are there any options/algorithm optimizations I could consider for further improving the performance?

**Comment by manuel-di-iorio** [comment]
With such high requirements I guess you need to look at fastest languages, writing the report in C++/Rust/Golang would be much more efficient, both in memory terms and performance
